examples/zero_copy/c++/zero_copy_subscriber.cxx

```diff
@@ -0,0 +1,497 @@
+    {
+        try {
+            DDS::PublicationBuiltinTopicData publication_data;
+            if (reader->get_matched_publication_data(
+                    publication_data,
+                    status.last_publication_handle) != DDS_RETCODE_OK) {
+                return;
+            }
+
+            int key = lookup_shmem_key_property(publication_data.property);
+
+            if (key == -1) {
+                return;
+            }
+
+            // Create the frame_set_view with the key and size. This frame set
+            // view attaches to the shared memory segment created by the frame
+            // set managed by the matched DataWriter
+            FrameSetView *frame_set_view = new FrameSetView(key, arg.size);
+            DDS::GUID_t publication_guid;
+            DDS_BuiltinTopicKey_to_guid(
+                    &publication_data.key,
+                    &publication_guid);
+
+            frame_set_views[publication_guid] = frame_set_view;
+        } catch (...) {
+            std::cout << "Exception in on_subscription_matched. " << std::endl;
+        }
+
+    }
+};
+
+class ZeroCopyListener : public ShmemKeyTrackingReaderListener {
+  public:
+    virtual void on_requested_deadline_missed(
+        DataReader* /*reader*/,
+        const RequestedDeadlineMissedStatus& /*status*/) {}
+
+    virtual void on_requested_incompatible_qos(
+        DataReader* /*reader*/,
+        const RequestedIncompatibleQosStatus& /*status*/) {}
+
+    virtual void on_sample_rejected(
+        DataReader* /*reader*/,
+        const SampleRejectedStatus& /*status*/) {}
+
+    virtual void on_liveliness_changed(
+        DataReader* /*reader*/,
+        const LivelinessChangedStatus& /*status*/) {}
+
+    virtual void on_sample_lost(
+        DataReader* /*reader*/,
+        const SampleLostStatus& /*status*/) {}
+
+    virtual void on_data_available(DataReader* reader);
+
+    ZeroCopyListener() :
+        total_latency(DDS_DURATION_ZERO),
+        received_frames(0),
+        shutdown_application(false)
+    {
+        clock = RTIHighResolutionClock_new();
+    }
+    ~ZeroCopyListener()
+    {
+        RTIHighResolutionClock_delete(clock);
+    }
+    bool start_shutdown()
+    {
+        return shutdown_application;
+    }
+  private:
+    // Sum of frame latency
+    DDS_Duration_t total_latency;
+    // Total number of received frames
+    int received_frames;
+    // Used for getting system time
+    RTIClock *clock;
+    struct RTINtpTime previous_print_time;
+    bool shutdown_application;
+
+    // Prints average latency in micro seconds
+    void print_average_latency()
+    {
+        std::cout << "Average latency: "
+                << ((total_latency.sec * 1000000)
+                        + (total_latency.nanosec / 1000)) / received_frames
+                << " microseconds" << std::endl;
+    }
+};
+
+void ZeroCopyListener::on_data_available(DataReader* reader)
+{
+    ZeroCopyDataReader *ZeroCopy_reader = NULL;
```

Comment:
I know a lot of them are from generated code, but don't follow the same (incorrect) pattern for the rest.